################################################################################
SHORT DESCRIPTION: 
################################################################################
Some miscellaneous fixes to make QEMU behave in OpenXT.

################################################################################
LONG DESCRIPTION: 
################################################################################
Fixes:
 - Don't propagate error when no VGA BIOS binary is found. This will cause
the VGA device creation to fail.
 - Have to still use absolute paths to the device-model node in XS when
setting the running state.
 - Set Xen PCI subsystem vendor and device IDs for VGA device.

################################################################################
CHANGELOG 
################################################################################
Author: Ross Philipson, philipsonr@ainfosec.com, 10/17/2016

################################################################################
REMOVAL 
################################################################################

################################################################################
UPSTREAM PLAN
################################################################################
No

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES 
################################################################################
Index: qemu-2.6.2/hw/pci/pci.c
===================================================================
--- qemu-2.6.2.orig/hw/pci/pci.c
+++ qemu-2.6.2/hw/pci/pci.c
@@ -2082,11 +2082,12 @@ static void pci_add_option_rom(PCIDevice
 
     size = get_image_size(path);
     if (size < 0) {
-        error_setg(errp, "failed to find romfile \"%s\"", pdev->romfile);
+        /* Do not propagate the error and cause device failure */
+        /* error_setg(errp, "failed to find romfile \"%s\"", pdev->romfile); */
         g_free(path);
         return;
     } else if (size == 0) {
-        error_setg(errp, "romfile \"%s\" is empty", pdev->romfile);
+        /*error_setg(errp, "romfile \"%s\" is empty", pdev->romfile); */
         g_free(path);
         return;
     }
Index: qemu-2.6.2/xen-common.c
===================================================================
--- qemu-2.6.2.orig/xen-common.c
+++ qemu-2.6.2/xen-common.c
@@ -95,7 +95,8 @@ static void xenstore_record_dm_state(str
         exit(1);
     }
 
-    snprintf(path, sizeof (path), "device-model/%u/state", xen_domid);
+    /* OpenXT: still using absolute paths so this was unhappy in a stubdom */
+    snprintf(path, sizeof (path), "/local/domain/0/device-model/%u/state", xen_domid);
     if (!xs_write(xs, XBT_NULL, path, state, strlen(state))) {
         fprintf(stderr, "error recording dm state\n");
         exit(1);
Index: qemu-2.6.2/hw/display/vga-pci.c
===================================================================
--- qemu-2.6.2.orig/hw/display/vga-pci.c
+++ qemu-2.6.2/hw/display/vga-pci.c
@@ -328,6 +328,8 @@ static void vga_pci_class_init(ObjectCla
 
     k->vendor_id = PCI_VENDOR_ID_QEMU;
     k->device_id = PCI_DEVICE_ID_QEMU_VGA;
+    k->subsystem_vendor_id = PCI_SUBVENDOR_ID_XEN;
+    k->subsystem_id = PCI_SUBDEVICE_ID_XEN;
     dc->vmsd = &vmstate_vga_pci;
     set_bit(DEVICE_CATEGORY_DISPLAY, dc->categories);
 }
Index: qemu-2.6.2/include/hw/pci/pci.h
===================================================================
--- qemu-2.6.2.orig/include/hw/pci/pci.h
+++ qemu-2.6.2/include/hw/pci/pci.h
@@ -71,6 +71,10 @@
 #define PCI_SUBVENDOR_ID_REDHAT_QUMRANET 0x1af4
 #define PCI_SUBDEVICE_ID_QEMU            0x1100
 
+/* Xen subsystem vendor ID and device */
+#define PCI_SUBVENDOR_ID_XEN             0x5853
+#define PCI_SUBDEVICE_ID_XEN             0x0001
+
 #define PCI_DEVICE_ID_VIRTIO_NET         0x1000
 #define PCI_DEVICE_ID_VIRTIO_BLOCK       0x1001
 #define PCI_DEVICE_ID_VIRTIO_BALLOON     0x1002
